connect/runtime/src/main/java/org/apache/kafka/connect/util/ConvertingFutureCallback.java

```diff
@@ -38,21 +44,42 @@ public ConvertingFutureCallback(Callback<T> underlying) {
 
     @Override
     public void onCompletion(Throwable error, U result) {
-        this.exception = error;
-        this.result = convert(result);
-        if (underlying != null)
-            underlying.onCompletion(error, this.result);
-        finishedLatch.countDown();
+        synchronized (this) {
+            if (isDone()) {
+                return;
+            }
+            
+            if (error != null) {
+                this.exception = error;
+            } else {
+                this.result = convert(result);
+            }
+
+            if (underlying != null)
+                underlying.onCompletion(error, this.result);
+            finishedLatch.countDown();
+        }
     }
 
     @Override
     public boolean cancel(boolean b) {
-        return false;
+        if (!b) {
```

Comment:
This implementation doesn't strictly adhere to the Future#cancel() contract - https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Future.html#cancel-boolean-

`After this method returns, subsequent calls to isDone() will always return true`. Before, we were ignoring the `mayInterruptIfRunning` flag and always cancelling when `cancel()` is called. I think I'd prefer that behavior than not actually cancelling when `cancel(false)` is called.

Reply:
> Before, we were ignoring the mayInterruptIfRunning flag and always cancelling when cancel() is called

I don't believe this is correct. Before, `cancel` was just a no-op and _no_ cancellation occurred, regardless of the value of `mayInterruptIfRunning`. That also violated the contract of the `Future` interface; I don't see this as a huge issue since we know that no other parts of the code base rely on the `cancel` method anyways.
Still, after reading the javadocs more carefully, I think it should be possible to implement `cancel` correctly if `mayInterruptIfRunning` is `false`.

Reply:
I was referring to a previous snapshot where you didn't check the flag.